### PR TITLE
Fixes to BotFrameworkAdapter for Proactive Messaging

### PR DESCRIPTION
**Pull request by cleemullins**
Needed to enable proactive messaging. Planning a more thorough fix in the near future, but this is needed to unblock customer.

### CODE REVIEW

**Comment by cleemullins** [comment]
@gabog @drub0y

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 1]
```diff
@@ -67,28 +82,40 @@ public new BotFrameworkAdapter Use(IMiddleware middleware)
...
             var connectorClient = new ConnectorClient(new Uri(activity.ServiceUrl), appCredentials);
-            return await connectorClient.Conversations.UpdateActivityAsync((Activity)activity);
+            return await connectorClient.Conversations.UpdateActivityAsync(activity);
         }
 
         public override async Task DeleteActivity(IBotContext context, ConversationReference reference)
         {
-            MicrosoftAppCredentials appCredentials = await GetAppCredentials((context as BotFrameworkBotContext).BotAppId);
+            BotFrameworkBotContext bfContext = context as BotFrameworkBotContext;
+            if (bfContext == null)
```
Comment (RamjotSingh) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Worth moving this to a separate method? This same code is being repeated at multiple places.

**Comment by RamjotSingh** [comment]
We also need to bring back EditActivity and DeleteActivity (preferably to all Adapters) as that functionality is currently hidden away behind ConnectorClient which is not visible directly.

**Comment by RamjotSingh** [comment]
Interestingly Update and Delete activity are not really callable right now since there is no way to get back the ActivityId from PostActivity. Are you planning work on that?

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 1 above)
Comment (cleemullins) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Yea, I'm actually hoping to change that today. 

Node already has it, and I wanted to get the "Fire and Forget" version(s) working first as the initial M2 changeset was so large I figured I would just do it after the main codebase was updated.

**Comment by cleemullins** [comment]
@RamjotSingh  Updated code per your comment to have a single method that validated the type of the IBotContext.

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs (hunk 1 above)
Comment (RamjotSingh) [reply libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: 👍

File: libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs [hunk 2]
```diff
@@ -61,34 +76,44 @@ public new BotFrameworkAdapter Use(IMiddleware middleware)
 
         public override async Task SendActivity(IBotContext context, params Activity[] activities)
         {
+            AssertBotFrameworkContext (context);
+
+            BotFrameworkBotContext bfContext = context as BotFrameworkBotContext;
+            MicrosoftAppCredentials appCredentials = await GetAppCredentials(bfContext.BotAppId);
```
Comment (RamjotSingh) [inline libraries/Microsoft.Bot.Builder/BotFrameworkAdapter.cs]: Hmm, For delay activities this is not really required. Though it probably wont really have a perf impact (Dictionary access and all that) but for clarity probably worth moving it inside the if condition. Assert can be left outside though.

**Pull request merged**